sympy/functions/elementary/piecewise.py

```diff
@@ -1029,6 +1029,83 @@ def _eval_rewrite_as_ITE(self, *args, **kwargs):
             last = ITE(c, a, last)
         return _canonical(last)
 
+    def _eval_rewrite_as_KroneckerDelta(self, *args):
+        from sympy import Ne, Eq, KroneckerDelta
+
+        conditions = []
+        true_value = None
+        for i in args:
+            if isinstance(i[1], Eq) or isinstance(i[1], Ne):
+                conditions.append(i)
+            elif isinstance(i[1], And) or isinstance(i[1], Or):
+                conditions.append(i)
+            elif i[1] is S.true:
+                if true_value is None:
+                    true_value = i[0]
+            else:
+                return
+
+        def rewrite_eq(args):
+            return KroneckerDelta(*args)
+
+        def rewrite_ne(args):
+            return KroneckerDelta(*args)
+
+        def rewrite_and(args):
+            k = 1
+            for c in args:
+                if isinstance(c, Eq):
+                    k *= rewrite_eq(c.args)
+                elif isinstance(c, Ne):
+                    k *= 1 - rewrite_ne(c.args)
+                elif isinstance(c, And):
+                    k *= rewrite_and(c.args)
+                elif isinstance(c, Or):
+                    k *= rewrite_or(c.args)
+                else:
+                    raise ValueError
+
+            return k
+
+        def rewrite_or(args):
+            k = 1
+            for c in args:
+                if isinstance(c, Eq):
+                    k *= 1 - rewrite_eq(c.args)
+                elif isinstance(c, Ne):
+                    k *= rewrite_ne(c.args)
+                elif isinstance(c, And):
+                    k *= rewrite_and(c.args)
+                elif isinstance(c, Or):
+                    k *= rewrite_or(c.args)
+                else:
+                    raise ValueError
+
+            return 1 - k
+
+        if true_value is not None:
+            conditions = conditions[::-1]
+            result = true_value
+
+            for i in conditions:
+                try:
+                    if isinstance(i[1], Eq):
+                        k = rewrite_eq(i[1].args)
+                    elif isinstance(i[1], Ne):
+                        k = 1 - rewrite_ne(i[1].args)
+                    elif isinstance(i[1], And):
+                        k = rewrite_and(i[1].args)
+                    elif isinstance(i[1], Or):
+                        k = rewrite_or(i[1].args)
+                    else:
+                        return
+                except ValueError:
+                    return
```

Comment:
Can we use a special exception class (perhaps just defined inside this function) instead of ValueError. I wouldn't want to catch a ValueError from anything else here.